internal/storage/serde.go

```
@@ -0,0 +1,414 @@
+		if err != nil {
+			return err
+		}
+		crr.rrs[i] = rr
+		crr.closers[i] = func() {
+			rr.Release()
+			er.Close()
+			reader.Close()
+		}
+	}
+	return nil
+}
+
+func (crr *compositeRecordReader) Next() error {
+	if crr.rrs == nil {
+		if crr.blobs == nil || len(crr.blobs) == 0 {
+			return io.EOF
+		}
+		crr.rrs = make([]array.RecordReader, len(crr.blobs))
+		crr.closers = make([]func(), len(crr.blobs))
+		crr.blobPos = -1
+		crr.fields = make([]FieldID, len(crr.rrs))
+		crr.r = compositeRecord{
+			recs:   make(map[FieldID]arrow.Record, len(crr.rrs)),
+			schema: make(map[FieldID]schemapb.DataType, len(crr.rrs)),
+		}
+		crr.iterateNextBatch()
+	}
+
+	composeRecord := func() bool {
+		for i, rr := range crr.rrs {
+			if ok := rr.Next(); !ok {
+				return false
+			}
+			// compose record
+			crr.r.recs[crr.fields[i]] = rr.Record()
+		}
+		return true
+	}
+
+	// Try compose records
+	if ok := composeRecord(); !ok {
+		// If failed the first time, try iterate next batch (blob), the error may be io.EOF
+		if err := crr.iterateNextBatch(); err != nil {
+			return err
+		}
+		// If iterate next batch success, try compose again
+		if ok := composeRecord(); !ok {
+			// If the next blob is empty, return io.EOF (it's rare).
+			return io.EOF
+		}
+	}
+	return nil
+}
+
+func (crr *compositeRecordReader) Record() Record {
+	return &crr.r
+}
+
+func (crr *compositeRecordReader) Close() {
+	for _, close := range crr.closers {
+		close()
+	}
+}
+
+func getBlobColLog(blob *Blob) (FieldID, UniqueID) {
+	// Blob key format (ref: data_codec.go):
+	// ${tenant}/insert_log/${collection_id}/${partition_id}/${segment_id}/${field_id}/${log_idx}
+	parts := strings.Split(blob.Key, "/")
+
+	id, _ := strconv.ParseInt(parts[len(parts)-1], 0, 10)
+	if len(parts) >= 2 {
+		colId, _ := strconv.ParseInt(parts[len(parts)-2], 0, 10)
+		return colId, id
+	}
+	return id, 0
+}
+
+func newCompositeRecordReader(blobs []*Blob) (*compositeRecordReader, error) {
```

maybe we should return an error if the length of splitted blob.Key is less than 2?